File: at_coder/abc30d.java
import java.util.Scanner;
import java.math.BigInteger;
import java.util.Arrays;

class Node{

    int value;
    int parent;

    public Node(int value, int parent){
        this.value = value;
        this.parent = parent;
    }
}


public class abc30d {
    public static void main(String[] args) {
        StragneDictionary sd = new StragneDictionary();
        sd.exec();
    }
}



class StragneDictionary {
    
    int N;
    int A;
    BigInteger K;
    Node[] board;
    int startLoopPoint;
    int startLoopPointCount;
    int loopLength;
    
    public StragneDictionary() {
        Scanner cin = new Scanner(System.in);
        this.N = cin.nextInt();
        this.A = cin.nextInt() - 1;
        this.K = new BigInteger(cin.next());

        this.board = new Node[N];
        for (int i = 0; i < N ; i++) {
            board[i] = new Node(i, cin.nextInt() - 1);    
        }

    }
    
    void exec() {
        findLoop();

        int ans = this.A;

        if(K.compareTo( BigInteger.valueOf(startLoopPointCount) )  > 0) {
            K = K.subtract(BigInteger.valueOf(startLoopPointCount));
            K = K.remainder(BigInteger.valueOf(loopLength) );
            ans = startLoopPoint;
        }
        int sK =  K.intValue();

        for (int i = 0; i < sK ; ++i) {
            ans = board[ans].parent;
        }

        System.out.println(ans + 1);
    }

    void findLoop(){
        int count = 0;
        int cur = this.A;

        int hashTable[] = new int [100001];
        Arrays.fill(hashTable, -1);
        hashTable[cur] = 0;

        while(true){
            count++;
            cur = board[cur].parent;
            if(hashTable[cur] != -1) break;
            hashTable[cur] = count;
        }
        this.loopLength = count - hashTable[cur];
        this.startLoopPointCount = hashTable[cur];
        this.startLoopPoint = cur;
    }
    
}

File: at_coder/abc30-d.cc
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
#include <string>

#define ll unsigned long long

using namespace std;

template<typename A, size_t N, typename T>

void Fill(A (&array)[N], const T &val){
    std::fill( (T*)array, (T*)(array+N), val );
}

struct Node{
	int v;
	int p;
};

// ll pow(ll a, ll e){
// 	ll ans = 1;
// 	while(e > 0){
// 		if(e & 1){
// 			ans *= a; 
// 		}
// 		a *= a;
// 		e >>= 1;  
// 	}
// 	return ans;
// }


int main(){
	ll n, a, k;
	string s;
	vector<Node> b;
	cin >> n >> a;
	cin >> s;
	a--;

	for (int i = 0; i < n; ++i)
	{
		int p;
		cin >> p;
		p--;
		Node node = {i, p};
		b.push_back(node);
	}

	ll count = 0;
	ll cur = a;

	int hash_table[100001];
	Fill(hash_table, -1);
	hash_table[cur] = 0;
	while(true){
		count++;
		cur = b[cur].p;
		if(hash_table[cur]!= -1) break; 
		hash_table[cur] = count;
	}

	ll loop_count = count - hash_table[cur];

	ll ans = a;

	ll ten_pow = 1;

	bool flag = true;


	// for (int i = s.length() - 1; i >= 0; --i)
	// {
	// 	k += (s[i] - '0') * ten_pow;
	// 	ten_pow *= 10;
	// 	if(k > hash_table[cur]){
	// 		if(flag){
	// 			flag = false;
	// 			k -= hash_table[cur];
	// 			k %= loop_count;
	// 		}
	// 	}

	// 	if(!flag){
	// 		k %= loop_count;
	// 		ten_pow %= loop_count;
	// 	}

	// 	// ten_pow %= loop_count;
	// 	// k %= loop_count;
	// 	// if(k > hash_table[cur]) break; 
	// }



	if(k > hash_table[cur]){
		k -= hash_table[cur];
		k %= loop_count;
		ans = cur;
	}

	for (int i = 0; i < k; ++i)
	{
		ans = b[ans].p;
		// cout << b[ans].p + 1 << endl;
	}

	cout << ans + 1 << endl;

}
